### PR TITLE
gadget: resolve index ambiguity between OnDiskStructure and LaidOutStructure

### PR DESCRIPTION
**Pull request by anonymouse64**
We had an issue previously, where we have both LaidOutStructure.Index and then
also (because LaidOutStructureIndex is embedded into OnDiskStructure)
OnDiskStructure.Index, but the meaning/usage of these two indices are subtly
different. The former is 0-referenced, and refers to the order in which all the
structures are mentioned in the abstract gadget.yaml, while the latter is
1-referenced and refers only to all structures which physically exist and are
measurable from the partition table (this excludes specifically MBR and other
type: bare structures).

To resolve this ambiguity and easy confusion, introduce a specific
OnDiskStructure.DiskIndex as well as rename the LaidOutStructure.Index to 
YamlIndex to make clear what the indices are referring to. Additionally, when
creating an OnDiskStructure to prevent possible confusion, use a new constant
UnknownLaidOutStructureIndex to make even more clear that OnDiskStructure's 
should not have their LaidOutStructure.YamlIndex used and should instead be 
using the DiskIndex.

Also split out the creation of an OnDiskStructure from a disks.Partition into
a new helper function since this will need to be done in other places too.

Based on top of https://github.com/snapcore/snapd/pull/11207

Broken out from https://github.com/snapcore/snapd/pull/11230

### CODE REVIEW

File: gadget/install/partition_test.go [hunk 1]
```diff
@@ -432,11 +433,11 @@ func (s *partitionTestSuite) TestRemovePartitions(c *C) {
 				},
 
 				StartOffset: 1024*1024 + 1024*1024,
-				Index:       2,
+				YamlIndex:   gadget.UnknownLaidOutDiskIndex,
```
Comment (pedronis) [inline gadget/install/partition_test.go:436]: the name of the constant is strange now given DiskIndex existence and what the constant is used for/assigned to?

**Review by pedronis — COMMENTED** [review]
lgtm, one consideration

File: gadget/install/partition_test.go (hunk 1 above)
Comment (anonymouse64) [reply gadget/install/partition_test.go:436]: do you think we still need to use/set the YamlIndex to this constant at all actually? it should hopefully be straight forward that this the YamlIndex is not to be used on an OnDiskStructure when there is DiskIndex

File: gadget/ondisk.go [hunk 2]
```diff
@@ -35,6 +35,15 @@ type OnDiskStructure struct {
 	// Node identifies the device node of the block device.
 	Node string
 
+	// DiskIndex is the index of the structure on the disk - this should be
+	// used instead of "Index" for an OnDiskStructure, "Index" comes from the
```
Comment (bboozzoo) [inline gadget/ondisk.go:39]: The comment needs an update, we don't have Index anymore

**Review by bboozzoo — APPROVED** [review]

File: gadget/install/partition_test.go (hunk 1 above)
Comment (pedronis) [reply gadget/install/partition_test.go:436]: it's probably OK to leave it out, I still hope one day to remove/clean up the messiness of OnDiskStructure vs LaidOutStructure etc

Comment (anonymouse64) [reply gadget/install/partition_test.go]: ok, I removed this constant

File: gadget/ondisk.go (hunk 2 above)
Comment (anonymouse64) [reply gadget/ondisk.go]: good catch, updated

**Review by pedronis — APPROVED** [review]
thanks

**Pull request merged**